compiler/src/dotty/tools/dotc/transform/localopt/RemoveUnnecessaryNullChecks.scala

```diff
@@ -6,104 +6,118 @@ import core.Contexts.Context
+
+  def transformer(implicit ctx: Context): Tree => Tree = 
+    tree => replace.getOrElse(tree, tree)
+
+
+  def visitor(implicit ctx: Context): Tree => Unit =
+    tree => tree match {
+
+      case Block(stats, expr) => 
+        def visitStatements(stats: List[Tree]): Unit =
+          stats match {
+            case s :: tail =>
+              s match {
+                case t: ValDef if !isVar(t.symbol) => 
+                  if (isNotNull(t.rhs)) tail.foreach(flagAsNotNull(t.symbol, _))
+                  else if (isNull(t.rhs)) tail.foreach(flagAsNull(t.symbol, _))
+
+                case Apply(Select(id: Ident, _), _) if !isVar(id.symbol) => 
+                  tail.foreach(flagAsNotNull(id.symbol, _))
+
+                case _ =>
+              }
+              visitStatements(tail)
+            case Nil => 
+          }
+
+        visitStatements(stats :+ expr)
+
+      case If(cond, th, el) => 
+        def visitBranches(cond: Tree, th: Tree, el: Tree): Unit = {
+          cond match {
+            case cond @ Select(expr, _) if cond.symbol == defn.Boolean_! =>
+              visitBranches(expr, el, th) 
+            case cond @ Apply(Select(lhs, _), List(rhs)) if cond.symbol == defn.Boolean_&& =>
+              visitBranches(lhs, th, el)
+              visitBranches(rhs, th, el)
+            case expr => 
+              nullChecksInExpr(expr) match {
+                case Some(symbol) => 
+                  flagAsNull(symbol, th)
+                  flagAsNotNull(symbol, el)
+                case None =>
+
+              }
+          }
         }
-        if (checkNonNull(vd.rhs, vd.symbol))
-          initializedVals += vd.symbol
-      }
-    case t: Tree =>
-  }
+       
+        visitBranches(cond, th, el)
+
+      case _ =>
+    } 
+
+  // return nullchecked symbol in expr (if any)
+  private def nullChecksInExpr(expr: Tree)(implicit ctx: Context): Option[Symbol] =
+    expr match {
+      case t @ Apply(Select(id: Ident, op), List(rhs)) if (t.symbol == defn.Object_eq || op == nme.EQ) && !isVar(id.symbol) && isNull(rhs) => Some(id.symbol)
+      case t @ Apply(Select(lhs, op), List(id: Ident)) if (t.symbol == defn.Object_eq || op == nme.EQ) && !isVar(id.symbol) && isNull(lhs) => Some(id.symbol)
+      case _ => None
+    }
 
+  // return all sub-trees of tree containing a nullcheck for symbol
+  private def findNullChecksFor(symbol: Symbol, tree: Tree)(implicit ctx: Context) = 
+    tree.filterSubTrees(nullChecksInExpr(_) match { case Some(s) if (s == symbol) => true; case _ => false })
+
+  // fill replace using findNullChecksFor
+  private def flagAsNull(symbol: Symbol, tree: Tree)(implicit ctx: Context) = replace ++= findNullChecksFor(symbol, tree).map((_, Literal(Constant(true))))
+  private def flagAsNotNull(symbol: Symbol, tree: Tree)(implicit ctx: Context) = replace ++= findNullChecksFor(symbol, tree).map((_, Literal(Constant(false))))
+
+  private def isNotNull(tree: Tree)(implicit ctx: Context): Boolean =
```

Comment:
It's a detail but I find the names are slightly misleading. I would call these `isNeverNull` and `isAlwaysNull`, otherwise it sounds like either `isNotNull(t)` or `isNull(t)` would be true for any `t`.